proposed/2023/Add-dotnet-search.md

```diff
@@ -0,0 +1,98 @@
+|||
+
+#### **Option `--source`**
+
+This option will specify a list of sources to search from. If a source is not specified using this option, the sources in the `nuget.config` file will be used.
+
+#### **Option `-exact-match`**
+
+- This option will allow for users to be able to search and have only exact matches as an output.
+- For example if a user uses `dotnet package search NuGet.CommandLine`
+
+        >NuGet.CommandLine | 6.7.0 | Downloads: N/A
+        >NuGet.CommandLine.XPlat | 6.7.0 | Downloads: N/A
+        >NuGet.Commands | 6.7.0 | Downloads: N/A
+        >NuGet.exe | 3.4.3 | Downloads: N/A
+        >NuGet.Bootstrapper | 2.6.0 | Downloads: N/A
+        >CommandLineParser20 | 2.0.0 | Downloads: N/A
+        >NuGet.VerifyMicrosoftPackage | 1.0.0 | Downloads: N/A
+        >NuGet.for.MSBuild | 2.1.0 ...
+
+- Using ``dotnet package search NuGet.CommandLine --exact-match`` on the other side will have the following output
+
+         >NuGet.CommandLine | 6.7.0 | Downloads: N/A
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+This command will prepare a search parameter using the inputs of  `Search Term`, `--take`, and `--prerelease`.
+This parameter is then used prepare an API query to the specified source. Nuget.Protocol will be used to do this query. Then the result is printed accordingly. If `--exact-match` is specified, assuming the search API returns a sorted search result, the first result of the search will be compared with the search term and printed out accordingly.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+
+## Rationale and alternatives
+
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+In nuget.exe there is `nuget.exe search` command which does the same thing. However, customers would like this functionality to be available in dotnet.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+I believe `dotnet tool search` is a subset to `dotnet package search`, as a result we could deprecate it. If it is necessary we could add an option in `dotnet package search` to allow users to specify that they want to only search dotnet tool.
```

Comment:
`dotnet tool search` is not a subset of `dotnet package search`. `tool search` specifically searches only for packages of type `DotnetTool`, whereas `dotnet package search` and `nuget.exe search` does not send a `packageType` query filter. Even if a `--package-type` filter was added to `dotnet package search`, `dotnet tool search` would still be valuable to avoid needing to memorize the package type identifier, though that could be partially mitigated by putting a list of well-known package types in the command help message. `dotnet tool search` also avoid customers needing to know that tools are shippable as NuGet packages and customers running `dotnet tool -h` can easily discover there's a search command.  I think there's value in `dotnet tool search` existing, rather than being deprecated if there's an alternative.

Reply:
I agree. we shouldn't deprecate, but there's maybe a world where they're implemented by the same code.